src/allotropy/parsers/utils/pandas.py

```diff
@@ -0,0 +1,115 @@
+from allotropy.parsers.utils.iterables import get_first_not_none
+from allotropy.parsers.utils.values import (
+    assert_not_none,
+    str_to_bool,
+    try_float_or_nan,
+    try_non_nan_float_or_none,
+)
+
+
+def rm_df_columns(data: pd.DataFrame, pattern: str) -> pd.DataFrame:
+    return data.drop(
+        columns=[column for column in data.columns if re.match(pattern, column)]
+    )
+
+
+def df_to_series(
+    df: pd.DataFrame, msg: str, index: int | None = None
+) -> pd.Series[Any]:
+    n_rows, _ = df.shape
+    if index is None and n_rows == 1:
+        index = 0
+    if index is None or index >= n_rows:
+        raise AllotropeConversionError(msg)
+    return pd.Series(df.iloc[index], index=df.columns)
+
+
+def df_to_series_data(df: pd.DataFrame, msg: str) -> SeriesData:
+    return SeriesData(df_to_series(df, msg))
+
+
+def assert_df_column(df: pd.DataFrame, column: str) -> pd.Series[Any]:
+    df_column = df.get(column)
+    if df_column is None:
+        msg = f"Unable to find column '{column}'"
+        raise AllotropeConversionError(msg)
+    return pd.Series(df_column)
+
+
+def assert_not_empty_df(df: pd.DataFrame, msg: str) -> pd.DataFrame:
+    if df.empty:
+        raise AllotropeConversionError(msg)
+    return df
+
+
+T = TypeVar("T", bool, float, int, str)
+
+
+class SeriesData:
+    def __init__(self, series: pd.Series[Any]) -> None:
+        self.series = series
+
+    def __getitem__(
+        self, type_and_key: tuple[Callable[..., T], str | Iterable[str]]
+    ) -> T:
+        # Implements index operator
+        type_, key = type_and_key
+        return assert_not_none(self.get(type_, key), str(key))
+
+    # This overload tells typing that if default is "None" then get might return None
+    @overload
+    def get(
+        self,
+        type_: Callable[..., T],
+        key: Iterable[str],
+        default: Literal[None] = None,
+    ) -> T | None:
+        ...
+
+    # This overload tells typing that if default matches T, get will return T
+    @overload
+    def get(self, type_: Callable[..., T], key: Iterable[str], default: T) -> T:
+        ...
+
+    def get(
+        self,
+        type_: Callable[..., T],
+        key: Iterable[str],
+        default: T | None = None,
+    ) -> T | None:
+        # Get value from series, if series is an iterable get the first non-null value
+        raw_value = (
+            self.series.get(key)
+            if isinstance(key, str)
+            else get_first_not_none(self.series.get, list(key))
+        )
+        try:
+            # bool needs special handling to convert
+            if type_ is bool:
+                raw_value = "true" if str_to_bool(raw_value) else ""
+            value = None if raw_value is None else type_(raw_value)
+        except ValueError:
+            value = None
+        return default if value is None else value
+
+    # TODO(nstender): I can't figure out how to integrate these yet, leaving as "try_..."
```

Comment:
I am definitely missing something but couldn't these be handled using an extra param plus some logic in the `ValueError` block in the above `get`? or have a flag for `.get` to be able to raise `ValueError` and then another function that wraps this function and sets that flag for additional handling

Reply:
For `non_nan` - it's more philosophical - bool flags in function signatures are discouraged. I might just go with it, but I deferred until later.

For `try_float_or_nan` - the problem is a typing one, it should just be:

`self.get(float, key, InvalidJsonFloat.NaN)`

but typing of the `get` function does not allow for returning `InvalidJsonFloat` when `T = float`. I'm sure it's doable, but haven't figured it out yet.

Reply:
Aaand fixed. Thank you for pushing me 🙏 😄